File: paleo_workbench/resources/data_asset_registry.py
"""DataAssetRegistry: Deep module for data asset classification, scanning, preview, and export.

Centralizes data format classification, directory scanning, format spec registration,
preview parser building, and format export dispatches into a single deep module.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path
from typing import TYPE_CHECKING, Any, Callable

from paleo_workbench.project.models import ExportArtifact, ResourceItem
from paleo_workbench.resources.classifier import classify_path as _classify_path
from paleo_workbench.resources.scanner import scan_resources as _scan_resources

if TYPE_CHECKING:
    from paleo_workbench.resources.preview_parsers.models import PreviewResult
    from paleo_workbench.resources.preview_settings import PreviewSettings


@dataclass
class FormatSpec:
    """Single-point registration spec for a data asset format."""

    format_id: str
    extensions: set[str] = field(default_factory=set)
    resource_type: str = "unknown"
    status: str = "indexed"
    preview_parser: Callable[[ResourceItem | ExportArtifact, Any], Any] | None = None
    exporter: Callable[[Any, str, Path], bool] | None = None


class DataAssetRegistry:
    """Deep module managing format specs, path classification, scanning, previews, and exports."""

    def __init__(self) -> None:
        self._format_specs: dict[str, FormatSpec] = {}

    def register_format(self, spec: FormatSpec) -> None:
        """Register a single-point format specification."""
        self._format_specs[spec.format_id.lower()] = spec
        for ext in spec.extensions:
            self._format_specs[ext.lower()] = spec

    def classify_path(self, path: Path) -> tuple[str, str, str]:
        """Classify file path into (resource_type, format, status)."""
        ext = path.suffix.lower().lstrip(".")
        if ext in self._format_specs:
            spec = self._format_specs[ext]
            return spec.resource_type, ext, spec.status
        return _classify_path(path)

    def inspect(self, path: Path) -> tuple[str, str, str]:
        """Inspect and classify file path into (resource_type, format, status). Alias for classify_path."""
        return self.classify_path(path)

    def scan_directory(
        self,
        root: Path,
        project_path: Path | None = None,
        *,
        skip_checksum_over_bytes: int | None = None,
        max_workers: int | None = None,
    ) -> list[ResourceItem]:
        """Scan directory and return list of classified ResourceItems."""
        return _scan_resources(
            root,
            project_path,
            skip_checksum_over_bytes=skip_checksum_over_bytes,
            max_workers=max_workers,
            classify=self.classify_path,
        )

    def parse_preview(
        self,
        asset: ResourceItem | ExportArtifact,
        settings: PreviewSettings,
        project_root: str | Path | None = None,
    ) -> PreviewResult:
        """Parse preview representation for an asset."""
        from paleo_workbench.resources.preview_parsers.registry import default_registry

        fmt = asset.format.lower()
        if fmt in self._format_specs and self._format_specs[fmt].preview_parser is not None:
            return self._format_specs[fmt].preview_parser(asset, settings)  # type: ignore

        registry = default_registry()
        return registry.build_preview(asset, settings, project_root=project_root)

    def export(self, asset: Any, format_id: str, output_path: str | Path) -> bool:
        """Export asset to output path in specified format.

        A registered :class:`FormatSpec` exporter wins; otherwise dispatch
        through the shared converter table (the same one behind
        ``get_available_formats``) so every export path has one
        implementation. Raises :class:`ExportError` when the format has no
        converter (the previous lazy import referenced a name that does not
        exist and would have raised ImportError here).
        """
        from paleo_workbench.resources.exporters import (
            ExportError,
            get_available_formats,
        )

        fmt = (format_id or "").lower()
        spec = self._format_specs.get(fmt)
        if spec is not None and spec.exporter is not None:
            return bool(spec.exporter(asset, format_id, Path(output_path)))

        label = fmt.upper()
        for available, convert_fn in get_available_formats(asset):
            if available == label:
                convert_fn(Path(asset.path), Path(output_path))
                return True
        raise ExportError(f"没有可用于 {label} 的导出器: {getattr(asset, 'format', '?')}")


# ---------------------------------------------------------------------------
# Global Singleton Instance
# ---------------------------------------------------------------------------
data_asset_registry = DataAssetRegistry()

File: libs/ui_data_core/src/data_asset_registry.cpp
// CONV-36 — resources/data_asset_registry.py port. See header.

#include "pwb/ui_data_core/data_asset_registry.hpp"

#include <algorithm>
#include <cctype>

#include <pwb/domain/text.hpp>
#include <pwb/interchange/exporters.hpp>

namespace pwb::ui_data_core {

void DataAssetRegistry::register_format(FormatSpec spec) {
    format_specs_[domain::lower_ascii(spec.format_id)] = spec;
    for (const auto& ext : spec.extensions) {
        format_specs_[domain::lower_ascii(ext)] = spec;
    }
}

std::tuple<std::string, std::string, std::string>
DataAssetRegistry::classify_path(const fs::path& path) const {
    std::string ext = path.extension().string();
    if (!ext.empty() && ext.front() == '.') ext.erase(ext.begin());
    ext = domain::lower_ascii(ext);
    if (const auto it = format_specs_.find(ext); it != format_specs_.end()) {
        return {it->second.resource_type, ext, it->second.status};
    }
    const ingest::Classification cls =
        ingest::classify_path(path.generic_string());
    return {cls.type, cls.format, cls.status};
}

std::vector<ResourceItem> DataAssetRegistry::scan_directory(
    const fs::path& root, const fs::path* project_path,
    std::int64_t skip_checksum_over_bytes, int max_workers,
    int io_slots) const {
    const ClassifyFn classify = [this](const fs::path& p) {
        return classify_path(p);
    };
    return scan_resources(root, project_path, skip_checksum_over_bytes,
                          max_workers, io_slots, classify);
}

ingest::preview::PreviewResult DataAssetRegistry::parse_preview(
    const ResourceItem& asset,
    const ingest::preview::PreviewSettings& settings,
    const std::optional<std::string>& project_root) const {
    const std::string fmt = domain::lower_ascii(asset.format);
    if (const auto it = format_specs_.find(fmt);
        it != format_specs_.end() && it->second.preview_parser) {
        return it->second.preview_parser(asset, settings);
    }

    ingest::preview::ResourceRef ref;
    ref.id = asset.id;
    ref.name = asset.name;
    ref.path = asset.path;
    ref.type = asset.type;
    ref.format = asset.format;
    ref.status = asset.status;
    if (asset.checksum.has_value()) ref.checksum = *asset.checksum;
    return ingest::preview::build_preview(ref, settings, project_root);
}

bool DataAssetRegistry::export_asset(const ResourceItem& asset,
                                     std::string_view format_id,
                                     const fs::path& output_path) const {
    const std::string fmt = domain::lower_ascii(format_id);
    if (const auto it = format_specs_.find(fmt);
        it != format_specs_.end() && it->second.exporter) {
        return it->second.exporter(asset, format_id, output_path);
    }

    // Shared converter table (same one behind get_available_formats).
    const std::string label = [&] {
        std::string up(fmt);
        for (char& c : up)
            c = static_cast<char>(std::toupper(static_cast<unsigned char>(c)));
        return up;
    }();
    for (const auto& [available, convert] :
         interchange::get_available_formats(asset.format)) {
        if (available == label) {
            convert(fs::path(asset.path), output_path);
            return true;
        }
    }
    throw interchange::ExportError("没有可用于 " + label +
                                   " 的导出器: " + asset.format);
}

DataAssetRegistry& data_asset_registry() {
    static DataAssetRegistry instance;
    return instance;
}

}  // namespace pwb::ui_data_core
